tensorboard/data/server/reservoir.rs

```diff
@@ -0,0 +1,329 @@
+    pub fn new(capacity: usize) -> Self {
+        Self::with_control(capacity, ChaCha20Rng::seed_from_u64(0))
+    }
+}
+
+impl<T, C: ReservoirControl> StageReservoir<T, C> {
+    /// Offers a record to the reservoir. The reservoir will always include the latest record.
+    /// Other than the latest record, the records kept form a simple random sample of the stream
+    /// (or at least approximately so in the case of preemptions).
+    pub fn offer(&mut self, step: Step, v: T) {
+        self.seen += 1;
+        let dst = self.ctl.destination(self.seen);
+
+        // Didn't make the cut? Keep-last only.
+        if dst >= self.capacity {
+            if self.staged_items.pop().is_none() {
+                self.committed_steps.pop();
+            }
+            self.staged_items.push((step, v));
+            return;
+        }
+
+        // Room to spare? Add to end.
+        let num_stored = self.committed_steps.len() + self.staged_items.len();
+        if num_stored < self.capacity {
+            self.staged_items.push((step, v));
+            return;
+        }
+
+        // Otherwise, we're evicting.
+        if dst < self.committed_steps.len() {
+            self.committed_steps.remove(dst);
+        } else {
+            self.staged_items.remove(dst - self.committed_steps.len());
+        }
+        self.staged_items.push((step, v))
+    }
+
+    /// Creates a new reservoir with the specified capacity and reservoir control.
+    ///
+    /// This function does not allocate. Reservoir capacity is allocated as records are offered.
+    pub fn with_control(capacity: usize, ctl: C) -> Self {
+        Self {
+            committed_steps: Vec::new(),
+            staged_items: Vec::new(),
+            capacity,
+            ctl,
+            seen: 0,
+        }
+    }
+
+    /// Accesses a view of the currently staged items.
+    pub fn staged_items(&self) -> &[(Step, T)] {
+        &self.staged_items[..]
+    }
+
+    /// Commits pending changes from this reservoir into a commit view. The commit should be a
+    /// vector that starts empty and is modified only by calls to `commit`/`commit_map` on this
+    /// reservoir value.
+    pub fn commit(&mut self, head: &mut Vec<(Step, T)>) {
+        self.commit_map(head, |t| t)
+    }
+
+    /// Commits pending changes from this reservoir into a commit view, applying a mapping function
+    /// to each new value. This can be used to perform relatively expensive conversions or
+    /// enrichments only for records that are actually committed. The commit should be a vector
+    /// that starts empty and is modified only by calls to `commit`/`commit_map` on this reservoir
+    /// value.
+    pub fn commit_map<S, F: FnMut(T) -> S>(&mut self, head: &mut Vec<(Step, S)>, mut f: F) {
+        let mut keep_steps = self.committed_steps.iter().peekable();
+        head.retain(move |(s, _)| match keep_steps.peek() {
```

Comment:
Nice catch—it does borrowck without, so I’ll remove it. I probably
reflexively added it because we use `keep_steps` inside the closure, but
it’s not needed because `keep_steps: usize` is `Copy`, and also because
we only use `&mut keep_steps`, anyway (implicitly, via `.next()`).

I think the right mental model for closures is just to desugar them—

```rust
struct MyClosure {
    things_moved_into_it: u32,
    things_it_closes_over: &str,
}

impl MyClosure {
    fn call(%%%self, args: TArgs) -> TReturn {
        unimplemented!()
    }
}
```

—where `%%%self` is either `self` for a `FnOnce`, `&self` for a `Fn`, or
`&mut self` for a `FnMut`. Then the normal borrowck rules will guide you
to whether something’s legal. So even with `move ||`, it’d be illegal to
write something like

```rust
fn main() {
    let four = String::from("four");
    let mut v = vec!["jumped", "over", "the", "lazy", "dog"];
    v.retain(move |s| {  // this is a `FnMut`, per `Vec::retain`
        // Try to move `four` via closure: can't, because you may have
        // moved it on a previous iteration.
        let to_check = four;
        s.len() == to_check.len()
    });
    dbg!(v);
}
```

because this is analogous to “cannot move out of borrowed content”:

```rust
struct MyClosure {
    four: String,
}

impl MyClosure {
    fn call(&mut self, s: &str) -> bool {
        let to_check = self.four; // cannot move out of borrowed content
        s.len() == to_check.len()
    }
}
```

(i.e., moving _into_ the closure is fine; it’s moving _out_ of it that’s
the problem here).

`move` can still be useful outside of threads; e.g.:

```rust
fn printer(s: String) -> impl Fn() -> () {
    move || { // this `move` is required for the program to compile
        println!("you said: {:?}", s);
    }
}

fn main() {
    let mut printers: Vec<Box<dyn Fn() -> ()>> = Vec::new();
    printers.push(Box::new(printer("hello".to_string())));
    printers.push(Box::new(printer("world".to_string())));
    for p in printers {
        p();
        p();
    }
}
```

Here maybe it’s easier to see that `s: String` is moved into the closure
when it’s constructed.

Reply:
Okay, that makes sense, thanks for the explanation.